File: sdks/node/src/proxy-client.ts
import * as grpc from "@grpc/grpc-js";
import { ProxyServiceClient } from "../gen/coordin8/proxy";
import type { Template } from "./registry-client";

export interface ProxyHandle {
  proxyId: string;
  localPort: number;
}

export class ProxyClient {
  private readonly stub: ProxyServiceClient;

  constructor(channel: grpc.Channel) {
    this.stub = new ProxyServiceClient(
      "passthrough:///djinn",
      grpc.credentials.createInsecure(),
      { channelOverride: channel }
    );
  }

  /** Ask the Djinn to open a local TCP forwarding port for the given template. */
  open(template: Template): Promise<ProxyHandle> {
    return new Promise((resolve, reject) => {
      this.stub.open({ template }, (err, res) => {
        if (err || !res) return reject(err);
        resolve({ proxyId: res.proxyId, localPort: res.localPort });
      });
    });
  }

  /** Release a proxy on the Djinn. */
  release(proxyId: string): Promise<void> {
    return new Promise((resolve, reject) => {
      this.stub.release({ proxyId }, (err) => {
        if (err) return reject(err);
        resolve();
      });
    });
  }

  /**
   * One-liner: open a proxy, resolve the local address, and pass it to a
   * factory function that creates the client.
   *
   * The factory receives `"localhost:<port>"` — use it to create any gRPC
   * client with your own credentials and options.
   *
   * ```typescript
   * const greeter = await djinn.proxy().proxyClient(
   *   addr => new GreeterServiceClient(addr, grpc.credentials.createInsecure()),
   *   { interface: "Greeter" }
   * );
   * ```
   */
  async proxyClient<T>(
    factory: (address: string) => T,
    template: Template
  ): Promise<T> {
    const handle = await this.open(template);
    return factory(`localhost:${handle.localPort}`);
  }
}

File: sdks/java/src/main/java/io/coordin8/ProxyClient.java
package io.coordin8;

import coordin8.ProxyServiceGrpc;
import coordin8.Proxy.OpenRequest;
import coordin8.Proxy.ReleaseRequest;
import coordin8.Proxy.ProxyHandle;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;

import java.util.Map;
import java.util.function.Function;

public class ProxyClient {

    private final ProxyServiceGrpc.ProxyServiceBlockingStub stub;

    ProxyClient(ManagedChannel channel) {
        this.stub = ProxyServiceGrpc.newBlockingStub(channel);
    }

    /**
     * Ask the Djinn to open a local TCP forwarding port for the given template.
     * Call {@link ProxyHandleRecord#close()} when done.
     */
    public ProxyHandleRecord open(Map<String, String> template) {
        ProxyHandle handle = stub.open(OpenRequest.newBuilder()
                .putAllTemplate(template)
                .build());
        return new ProxyHandleRecord(handle.getProxyId(), handle.getLocalPort(), this);
    }

    /** Release a proxy on the Djinn. */
    public void release(String proxyId) {
        stub.release(ReleaseRequest.newBuilder().setProxyId(proxyId).build());
    }

    /**
     * One-liner: open a proxy, build a ManagedChannel to it, pass it to
     * the stub factory, and return the stub. The returned stub is ready to use.
     *
     * <p>The caller is responsible for shutting down the channel and closing
     * the proxy. For a managed lifecycle, use {@link #open} directly.
     *
     * <pre>{@code
     * var greeter = djinn.proxy().proxyStub(
     *     GreeterServiceGrpc::newBlockingStub,
     *     Map.of("interface", "Greeter")
     * );
     * }</pre>
     */
    public <T> T proxyStub(Function<ManagedChannel, T> factory, Map<String, String> template) {
        ProxyHandleRecord handle = open(template);
        ManagedChannel channel = ManagedChannelBuilder
                .forAddress("localhost", handle.localPort())
                .usePlaintext()
                .build();
        return factory.apply(channel);
    }

    public record ProxyHandleRecord(
            String proxyId,
            int localPort,
            ProxyClient client
    ) implements AutoCloseable {
        @Override
        public void close() {
            client.release(proxyId);
        }
    }
}
